# HG changeset patch
# User Jaroslav Hajek <highegg@gmail.com>
# Date 1275388456 -7200
#      Tue Jun 01 12:34:16 2010 +0200
# Node ID e3064439d6b487f8f6bfca35286620cbdb32e20a
# Parent  b17a966099edd2355fb63e3d9cad576dfaccbedb
new Array method for internal use

diff --git a/liboctave/Array.cc b/liboctave/Array.cc
--- a/liboctave/Array.cc
+++ b/liboctave/Array.cc
@@ -2566,16 +2566,26 @@ Array<T>::print_info (std::ostream& os, 
 
   // 2D info:
   //
   //     << pefix << "rows: " << rows () << "\n"
   //     << prefix << "cols: " << cols () << "\n";
 }
 
 template <class T>
+bool Array<T>::optimize_dimensions (const dim_vector& dv)
+{
+  bool retval = dimensions == dv;
+  if (retval)
+    dimensions = dv;
+  
+  return retval;
+}
+
+template <class T>
 void Array<T>::instantiation_guard ()
 {
   // This guards against accidental implicit instantiations.
   // Array<T> instances should always be explicit and use INSTANTIATE_ARRAY.
   T::__xXxXx__();
 }
 
 #define INSTANTIATE_ARRAY(T, API) \
diff --git a/liboctave/Array.h b/liboctave/Array.h
--- a/liboctave/Array.h
+++ b/liboctave/Array.h
@@ -617,16 +617,21 @@ public:
 
   template <class U>
   Array<U>
   map (U (&fcn) (const T&)) const
   { return map<U, U (&) (const T&)> (fcn); }
 
   template <class U> friend class Array;
 
+  // Returns true if this->dims () == dv, and if so, replaces this->dimensions
+  // by a shallow copy of dv. This is useful for maintaining several arrays with
+  // supposedly equal dimensions (e.g. structs in the interpreter).
+  bool optimize_dimensions (const dim_vector& dv);
+
 private:
   static void instantiation_guard ();
 };
 
 // This is a simple wrapper template that will subclass an Array<T> type or any
 // later type derived from it and override the default non-const operator() to
 // not check for the array's uniqueness. It is, however, the user's
 // responsibility to ensure the array is actually unaliased whenever elements
diff --git a/liboctave/ChangeLog b/liboctave/ChangeLog
--- a/liboctave/ChangeLog
+++ b/liboctave/ChangeLog
@@ -1,8 +1,13 @@
+2010-06-01  Jaroslav Hajek  <highegg@gmail.com>
+
+	* Array.cc (Array<T>::optimize_dimensions): New method.
+	* Array.h: Declare it.
+
 2010-06-01  Jaroslav Hajek  <highegg@gmail.com>
 
 	* Array.cc (Array<T>::maybe_delete_dims): Remove method (makes no
 	longer sense).
 	* Array.h: Update decls.
 
 2010-05-20  Jaroslav Hajek  <highegg@gmail.com>
 
